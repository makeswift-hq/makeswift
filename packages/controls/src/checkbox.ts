import { ControlDataTypeKey } from "./common/data-type-key"

export const CheckboxControlDataTypeKey = ControlDataTypeKey

export const CheckboxControlDataTypeValueV1 = 'checkbox::v1'

export type CheckboxControlDataV0 = boolean

export type CheckboxControlDataV1 = {
  [CheckboxControlDataTypeKey]: typeof CheckboxControlDataTypeValueV1
  value: boolean,
}

export type CheckboxControlData = CheckboxControlDataV0 | CheckboxControlDataV1

export const CheckboxControlType = 'makeswift::controls::checkbox'

type CheckboxControlConfig = {
  label?: string
  defaultValue?: boolean
}

export type CheckboxControlDefinition<C extends CheckboxControlConfig = CheckboxControlConfig> = {
  type: typeof CheckboxControlType
  config: C
  version?: 1
}

export function Checkbox<C extends CheckboxControlConfig>(
  config: C = {} as C,
): CheckboxControlDefinition<C> {
  return { type: CheckboxControlType, config, version: 1 }
}
