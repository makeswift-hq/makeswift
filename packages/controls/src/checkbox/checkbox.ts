import { match } from 'ts-pattern'
import { z } from 'zod'

import {
  ControlDataTypeKey,
  type ValueType as _ValueType,
  type CopyContext,
} from '../common'
import { type ResourceResolver } from '../resource-resolver'
import { controlTraitsRegistry } from '../registry'

import {
  type VersionedControlDefinition,
  type ControlTraits,
  type ParseResult,
  type ValueSubscription,
} from '../traits'

import {
  DefaultControlInstance,
  ControlInstance,
  type Send,
} from '../control-instance'

import { ControlDefinition, serializeConfig } from '../control-definition'
import { WithAssociatedTypes } from '../utils/associated-types'

export const Checkbox = controlTraitsRegistry.add(
  (() => {
    const type = 'makeswift::controls::checkbox' as const
    const v1DataType = 'checkbox::v1' as const
    const version = 1 as const

    const dataSignature = {
      v1: { [ControlDataTypeKey]: v1DataType },
    } as const

    const dataSchema = z.union([
      z.boolean(),
      z.object({
        [ControlDataTypeKey]: z.literal(v1DataType),
        value: z.boolean(),
      }),
    ])

    const configSchema = z.object({
      label: z.string().optional(),
      defaultValue: z.boolean().optional(),
    })

    type ControlData = z.infer<typeof dataSchema>
    type Config = z.infer<typeof configSchema>

    type ControlDefinition<C extends Config = Config> =
      VersionedControlDefinition<
        typeof type,
        C,
        undefined extends C['defaultValue'] ? boolean | undefined : boolean,
        typeof version
      >

    const ctor = <C extends Config>(config?: C): ControlDefinition<C> => ({
      type,
      config: config ?? ({} as C),
      version,
    })

    ctor.controlType = type
    ctor.dataSignature = dataSignature

    ctor.safeParse = (
      data: unknown | undefined,
    ): ParseResult<ControlData | undefined> => {
      const result = dataSchema.optional().safeParse(data)
      return result.success
        ? { success: true, data: result.data }
        : { success: false, error: result.error.flatten().formErrors[0] }
    }

    ctor.fromData = (
      data: ControlData | undefined,
      definition: ControlDefinition,
    ) => {
      return match(data)
        .with(dataSignature.v1, ({ value }) => value)
        .otherwise((value) => value ?? definition?.config?.defaultValue)
    }

    ctor.toData = (
      value: boolean,
      definition: ControlDefinition,
    ): ControlData => {
      return match('version' in definition ? definition.version : undefined)
        .with(version, () => ({
          ...dataSignature.v1,
          value,
        }))
        .with(undefined, () => value)
        .exhaustive()
    }

    ctor.resolveValue = (
      value: _ValueType<ControlDefinition>,
      _resolver: ResourceResolver,
      definition: ControlDefinition,
    ): Promise<boolean | undefined> => {
      return Promise.resolve(ctor.fromData(value, definition))
    }

    ctor.subscribeValue = (
      value: _ValueType<ControlDefinition>,
      definition: ControlDefinition,
      _resolver: ResourceResolver,
    ): ValueSubscription<boolean | undefined> => {
      return {
        readValue() {
          if (value === undefined) return definition.config.defaultValue
          return ctor.fromData(value, definition)
        },
        subscribe() {
          return () => {}
        },
      }
    }

    ctor.createInstance = (send: Send) => new DefaultControlInstance(send)

    ctor.getSwatchIds = (_data: ControlData | undefined): string[] => []

    return ctor as typeof ctor &
      ControlTraits<typeof type, ControlData, ControlDefinition>
  })(),
)

type DataType = z.infer<typeof Definition.schema.data>
type Config = z.infer<typeof Definition.schema.config>

type ValueType<C extends Config> = undefined extends C['defaultValue']
  ? boolean | undefined
  : boolean

class Definition<C extends Config = Config> extends ControlDefinition<
  C,
  DataType,
  ValueType<C>
> {
  private static readonly v1DataType = 'checkbox::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::checkbox' as const

  static get schema() {
    const value = z.boolean().optional()
    const data = z.union([
      value,
      z.object({
        [ControlDataTypeKey]: z.literal(this.v1DataType),
        value,
      }),
    ])

    const config = z.object({
      label: z.string().optional(),
      defaultValue: z.boolean().optional(),
    })

    const version = z.literal(1).optional()
    const definition = z.object({
      type: z.literal(this.type),
      config,
      version,
    })

    return { value, data, config, version, definition }
  }

  static deserialize(data: unknown): Definition {
    const { config, version } = this.schema.definition.parse(data)
    return new Definition(config, version)
  }

  constructor(
    readonly config: C,
    readonly version: z.infer<typeof Definition.schema.version> = 1,
  ) {
    super(config)
  }

  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
    const result = Definition.schema.data.optional().safeParse(data)
    return result.success
      ? { success: true, data: result.data }
      : { success: false, error: result.error.flatten().formErrors[0] }
  }

  fromData(data: DataType | undefined): ValueType<C> | undefined {
    return match(data)
      .with(Definition.dataSignature.v1, ({ value }) => value)
      .otherwise((value) => value ?? this.config.defaultValue)
  }

  toData(value: ValueType<C>): DataType {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType | undefined,
    _context: CopyContext,
  ): DataType | undefined {
    return data
  }

  resolveValue(
    _value: ValueType<C>,
    _resolver: ResourceResolver,
  ): Promise<void> {
    return Promise.resolve()
  }

  createInstance(send: Send): ControlInstance<unknown> {
    return new DefaultControlInstance(send)
  }

  serialize(): [unknown, Transferable[]] {
    const [config, transferables] = serializeConfig(this.config)
    return [
      {
        type: Definition.type,
        config,
        version: this.version,
      },
      transferables,
    ]
  }

  getSwatchIds(_data: DataType | undefined): string[] {
    return []
  }
}

export const CheckboxV2 = (() => {
  const ctor = <C extends Config>(config?: C): Definition<C> =>
    new Definition(config ?? ({} as C))

  ctor.controlType = Definition.type
  return ctor as typeof ctor &
    WithAssociatedTypes<{
      Definition: Definition
    }>
})()
