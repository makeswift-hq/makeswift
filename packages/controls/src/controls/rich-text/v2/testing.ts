import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
