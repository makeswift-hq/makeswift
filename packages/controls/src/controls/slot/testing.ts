import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
