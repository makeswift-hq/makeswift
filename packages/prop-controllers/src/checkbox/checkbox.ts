import { z } from 'zod'
import { ControlDataTypeKey, Options, Types } from '../prop-controllers'
import { P, match } from 'ts-pattern'

const checkboxPropControllerDataV0Schema = z.boolean()

export type CheckboxPropControllerDataV0 = z.infer<
  typeof checkboxPropControllerDataV0Schema
>

export const CheckboxPropControllerDataV1Type = 'prop-controllers::checkbox::v1'

const checkboxPropControllerDataV1Schema = z.object({
  [ControlDataTypeKey]: z.literal(CheckboxPropControllerDataV1Type),
  value: z.boolean(),
})

export type CheckboxPropControllerDataV1 = z.infer<
  typeof checkboxPropControllerDataV1Schema
>

export const checkboxPropControllerDataSchema = z.union([
  checkboxPropControllerDataV0Schema,
  checkboxPropControllerDataV1Schema,
])

export type CheckboxPropControllerData = z.infer<
  typeof checkboxPropControllerDataSchema
>

export type CheckboxOptions = Options<{
  preset?: CheckboxPropControllerData
  label: string
  hidden?: boolean
}>

type CheckboxDescriptorV0<_T = CheckboxPropControllerDataV0> = {
  type: typeof Types.Checkbox
  options: CheckboxOptions
}

type CheckboxDescriptorV1<
  _T = CheckboxPropControllerData,
  U extends CheckboxOptions = CheckboxOptions,
> = {
  type: typeof Types.Checkbox
  version: 1
  options: U
}

export type CheckboxDescriptor<_T = CheckboxPropControllerData> =
  | CheckboxDescriptorV0
  | CheckboxDescriptorV1

export type ResolveCheckboxPropControllerValue<T extends CheckboxDescriptor> =
  T extends CheckboxDescriptor ? boolean | undefined : never

/**
 * @deprecated Imports from @makeswift/prop-controllers are deprecated. Use
 * @makeswift/runtime/controls instead.
 */
export function Checkbox(options: CheckboxOptions): CheckboxDescriptorV1 {
  return { type: Types.Checkbox, version: 1, options }
}

export function getCheckboxPropControllerDataBoolean(
  data: CheckboxPropControllerData | undefined,
): boolean | undefined {
  return match(data)
    .with(
      { [ControlDataTypeKey]: CheckboxPropControllerDataV1Type },
      (v1) => v1.value,
    )
    .otherwise((v0) => v0)
}

export function createCheckboxPropControllerDataFromBoolean(
  value: boolean,
  definition?: CheckboxDescriptor,
): CheckboxPropControllerData {
  return match(definition)
    .with(
      { version: 1 },
      P.nullish,
      () =>
        ({
          [ControlDataTypeKey]: CheckboxPropControllerDataV1Type,
          value,
        } as const),
    )
    .otherwise(() => value)
}
