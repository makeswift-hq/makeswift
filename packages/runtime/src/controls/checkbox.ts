export type CheckboxControlData = boolean

export const CheckboxControlType = 'makeswift::controls::checkbox'

type CheckboxControlConfig = {
  label?: string
  defaultValue?: boolean
}

export type CheckboxControlDefinition<C extends CheckboxControlConfig = CheckboxControlConfig> = {
  type: typeof CheckboxControlType
  config: C
}

export function Checkbox<C extends CheckboxControlConfig>(
  config: C = {} as C,
): CheckboxControlDefinition<C> {
  return { type: CheckboxControlType, config }
}
