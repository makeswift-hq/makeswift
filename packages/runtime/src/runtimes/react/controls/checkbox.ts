import { type DataType, type ResolvedValueType } from '@makeswift/controls'

import { CheckboxDefinition } from '../../../controls'

export function useCheckboxControlValue(
  data: DataType<CheckboxDefinition> | undefined,
  definition: CheckboxDefinition,
): ResolvedValueType<CheckboxDefinition> {
  return definition.fromData(data) ?? definition.config.defaultValue
}
