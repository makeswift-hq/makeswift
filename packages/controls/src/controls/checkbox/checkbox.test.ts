import { testDefinition, testResolveValue } from '../../testing/test-definition'

import { Checkbox, CheckboxDefinition } from './checkbox'

describe('Checkbox', () => {
  describe('constructor', () => {
    test.each([true, false, undefined])(
      'call with default value `%s` returns versioned definition',
      (value) => {
        expect(
          Checkbox({
            label: 'visible',
            defaultValue: value,
          }),
        ).toMatchSnapshot()
      },
    )

    test('disallows extraneous properties', () => {
      Checkbox({
        label: undefined,
        defaultValue: undefined,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: CheckboxDefinition) {}
    assignTest(Checkbox())
    assignTest(Checkbox({ label: 'visible' }))
    assignTest(Checkbox({ defaultValue: true }))
    assignTest(Checkbox({ label: 'visible', defaultValue: true }))
    assignTest(Checkbox({ defaultValue: false as boolean }))
    assignTest(Checkbox({ defaultValue: undefined }))
    assignTest(Checkbox({ label: 'visible', defaultValue: undefined }))
    assignTest(Checkbox({ label: undefined, defaultValue: undefined }))
  })
})

describe.each([
  [Checkbox({ defaultValue: false, label: 'visible' }), [true, false]],
  [Checkbox({ label: 'disabled' }), [true, false, undefined]],
])('Checkbox', (def, values) => {
  const invalidValues = [null, 17, 'text']
  testDefinition(def, values, invalidValues)
  testResolveValue(def, values)
})
