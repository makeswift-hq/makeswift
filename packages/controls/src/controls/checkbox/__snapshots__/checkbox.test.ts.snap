// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox constructor call with default value \`false\` returns versioned definition 1`] = `
CheckboxDefinition {
  "config": {
    "defaultValue": false,
    "label": "visible",
  },
  "version": 1,
}
`;

exports[`Checkbox constructor call with default value \`true\` returns versioned definition 1`] = `
CheckboxDefinition {
  "config": {
    "defaultValue": true,
    "label": "visible",
  },
  "version": 1,
}
`;

exports[`Checkbox constructor call with default value \`undefined\` returns versioned definition 1`] = `
CheckboxDefinition {
  "config": {
    "defaultValue": undefined,
    "label": "visible",
  },
  "version": 1,
}
`;

exports[`Checkbox definition w/ config {"defaultValue":false,"label":"visible"} toData returns v1 data for \`false\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "checkbox::v1",
  "value": false,
}
`;

exports[`Checkbox definition w/ config {"defaultValue":false,"label":"visible"} toData returns v1 data for \`true\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "checkbox::v1",
  "value": true,
}
`;

exports[`Checkbox definition w/ config {"label":"disabled"} toData returns v1 data for \`false\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "checkbox::v1",
  "value": false,
}
`;

exports[`Checkbox definition w/ config {"label":"disabled"} toData returns v1 data for \`true\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "checkbox::v1",
  "value": true,
}
`;

exports[`Checkbox definition w/ config {"label":"disabled"} toData returns v1 data for \`undefined\` when definition version is 1 1`] = `undefined`;
